src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,527 @@
+import "./AllocationManagerStorage.sol";
+
+contract AllocationManager is
+    Initializable,
+    OwnableUpgradeable,
+    Pausable,
+    AllocationManagerStorage,
+    ReentrancyGuardUpgradeable
+{
+    using Snapshots for Snapshots.DefaultWadHistory;
+    using DoubleEndedQueue for DoubleEndedQueue.Bytes32Deque;
+    using SlashingLib for uint256;
+
+    /**
+     *
+     *                         INITIALIZING FUNCTIONS
+     *
+     */
+
+    /**
+     * @dev Initializes the immutable addresses of the strategy mananger, delegationManage,
+     * and eigenpodManager contracts
+     */
+    constructor(
+        IDelegationManager _delegation,
+        IAVSDirectory _avsDirectory,
+        uint32 _DEALLOCATION_DELAY,
+        uint32 _ALLOCATION_CONFIGURATION_DELAY
+    ) AllocationManagerStorage(_delegation, _avsDirectory, _DEALLOCATION_DELAY, _ALLOCATION_CONFIGURATION_DELAY) {
+        _disableInitializers();
+    }
+
+    /// @inheritdoc IAllocationManager
+    function initialize(
+        address initialOwner,
+        IPauserRegistry _pauserRegistry,
+        uint256 initialPausedStatus
+    ) external initializer {
+        _initializePauser(_pauserRegistry, initialPausedStatus);
+        _transferOwnership(initialOwner);
+    }
+
+    /// @inheritdoc IAllocationManager
+    function slashOperator(
+        SlashingParams calldata params
+    ) external onlyWhenNotPaused(PAUSED_OPERATOR_SLASHING) {
+        require(0 < params.wadToSlash && params.wadToSlash <= WAD, InvalidWadToSlash());
+
+        // Check that the operator is registered and slashable
+        OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: params.operatorSetId});
+        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
+        require(avsDirectory.isOperatorSlashable(params.operator, operatorSet), InvalidOperator());
+
+        // Record the proportion of 1e18 that the operator's total shares that are being slashed
+        uint256[] memory wadSlashed = new uint256[](params.strategies.length);
+
+        for (uint256 i = 0; i < params.strategies.length; ++i) {
+            PendingMagnitudeInfo memory info =
+                _getPendingMagnitudeInfo(params.operator, params.strategies[i], operatorSetKey);
+
+            require(info.currentMagnitude > 0, OperatorNotAllocated());
+
+            // 1. Calculate slashing amount and update current/encumbered magnitude
+            uint64 slashedMagnitude = uint64(uint256(info.currentMagnitude).mulWad(params.wadToSlash));
+            info.currentMagnitude -= slashedMagnitude;
+            info.encumberedMagnitude -= slashedMagnitude;
+
+            // 2. If there is a pending deallocation, reduce pending deallocation proportionally.
+            // This ensures that when the deallocation is completed, less magnitude is freed.
+            if (info.pendingDiff < 0) {
+                uint64 slashedPending = uint64(uint256(uint128(-info.pendingDiff)).mulWad(params.wadToSlash));
```

Comment:
`slashedPending` is not used anywhere else so can we remove this uint64 cast?

Reply:
Ahh based on my comment below, this amount is supposed to be added to `slashedMagnitude` but that line is missing.